### ISSUE TITLE
Calling BulkInsertOrUpdate with SetOutputIdentity=true, PreserveInsertOrder=true, returns data that contains data from other rows.

### ISSUE THREAD

**Issue by MiladNL** [borisdj/EFCore.BulkExtensions#1629]
I have a table of instances. Example:

id:1, someValue: A
id:0, someValue: B
id:0, someValue: C

Instances contains database records with some of them Id = 0 (to be inserted) and some records id > 0 (to be updated).
Column id is an Identity Column:

![image](https://github.com/user-attachments/assets/4c0408ec-3fee-427c-a07c-498870082a4c)

` var bulkConfig = new BulkConfig
  {
      SetOutputIdentity = true,
      PreserveInsertOrder = true,
      SRID = Constants.DefaultSrid
  };
    _context.BulkInsertOrUpdate(instances, bulkConfig);`

After running the code, instances will be populated with data from other rows:
id:1, someValue: C
id:2, someValue: A
id:3, someValue: B

This happens in EFCore.BulkExtensions 8.1.2 (8.1.1 is working correctly).

**Comment by streamBASETdS** [comment]
Hi, I can confirm that the IDs returned by the BulkInsert call when PreserveInsertOrder and SetOutputIdentity is true are arbitrarily sorted and NOT correct. Entity instances get different ID values than is written to the DB. If you use those returned IDs to set FKs, it results in absolute chaos  :-(

**Comment by buituansonHE172207** [comment]
I have the same issue and i have to change to use add range async

**Comment by borisdj** [comment]
Can you make a Test for this issue.
I have added this one and it passed:
```C#
[Theory]
[InlineData(SqlType.SqlServer, true)]
public void BulkOrderedInsertTest(SqlType sqlType)
{
  ContextUtil.DatabaseType = sqlType;
  using var context = new TestContext(ContextUtil.GetOptions());

  context.Items.Add(new Item { ItemId = 0, Name = "Name 0", TimeUpdated = DateTime.Now });
  context.SaveChanges();

  var entities = new List<Item>();
  entities.Add(new Item { ItemId = 1, Name = "Name A", TimeUpdated = DateTime.Now,});
  entities.Add(new Item { ItemId = 0, Name = "Name B", TimeUpdated = DateTime.Now, });
  entities.Add(new Item { ItemId = 0, Name = "Name C", TimeUpdated = DateTime.Now, });

  var bulkConfig = new BulkConfig
  {
    SetOutputIdentity = true,
    PreserveInsertOrder = true
  };
  context.BulkInsertOrUpdate(entities, bulkConfig);

  Assert.Equal(1, entities[0].ItemId);
  Assert.Equal(2, entities[1].ItemId);
  Assert.Equal(3, entities[2].ItemId);

  Assert.Equal("Name A", entities[0].Name);
  Assert.Equal("Name B", entities[1].Name);
  Assert.Equal("Name C", entities[2].Name);
}
```

**Comment by buituansonHE172207** [comment]
I think you should test with auto identity with guid and using bulk insert, i have problem when using this
![image](https://github.com/user-attachments/assets/795ccd74-a685-4271-825b-749f106380b6)
![image](https://github.com/user-attachments/assets/a7a7a4bf-28c5-45d4-985f-4ecc457a8605)

**Comment by streamBASETdS** [comment]
> Can you make a Test for this issue. I have added this one and it passed:
> 

For me, the issue doesn't occur for small sized bulk inserts (a few entries). When bulkinserting larger numbers of entities (1000+), the issue shows. I'm just inserting new entities (not upserting); and I'm using the technique to assign incrementing "pre-IDs" to define the sort order. I'm using negative numbers for this to differentiate between new and existing entities when debugging. Up to 8.1.1 this works perfectly.

**Comment by streamBASETdS** [comment]
No, I meant I use negative "pre-IDs" for new entries, and I use them also to differentiate during debugging (so if I encounter entities with negative IDs while stepping through code, I know those are new ones that will be bulk-inserted later; positive IDs are existing entities that are not bulked whatsoever but rather updated in a completely different codepath that allows for dynamic diffing/historising of data etc). Everything that is fed to EFCore.BulkExtensions is "new" with negative IDs.

**Comment by borisdj** [comment]
@streamBASETdS are you using SqlServer, here is a new sample and it also passes:
```C#
[Theory]
[InlineData(SqlType.SqlServer)]
public void BulkOrderedInsertTest(SqlType sqlType)
{
  ContextUtil.DatabaseType = sqlType;
  using var context = new TestContext(ContextUtil.GetOptions());

  context.Items.RemoveRange(context.Items.ToList());
  context.SaveChanges();

  int counter0 = 1_000;
  var entities0 = new List<Item>();
  for (int i = 1; i <= counter0; i++)
  {
    entities0.Add(new Item { Name = "Name Init" + i, TimeUpdated = DateTime.Now });
  }
  context.Items.AddRange(entities0);
  context.SaveChanges();

  int startingIdNumber = entities0[counter0 - 1].ItemId;

  var entities = new List<Item>();
  int counter = - 10_000;
  int number = 0;
  for (int i = counter; i < 0 ; i++)
  {
    number++;
    entities.Add(new Item { ItemId = i, Name = "Name Bulk " + number, TimeUpdated = DateTime.Now });
  }

  var bulkConfig = new BulkConfig
  {
    SetOutputIdentity = true,
    PreserveInsertOrder = true
  };
  context.BulkInsert(entities, bulkConfig);

  Assert.Equal(startingIdNumber + 0 + 1, entities[0].ItemId);
  Assert.Equal(startingIdNumber + 1000 + 1, entities[1000].ItemId);
  Assert.Equal(startingIdNumber + 2500 + 1, entities[2500].ItemId);
  Assert.Equal(startingIdNumber + 4678 + 1, entities[4678].ItemId);
  Assert.Equal(startingIdNumber + 8888 + 1, entities[8888].ItemId);

  Assert.Equal("Name Bulk " + (0 + 1), entities[0].Name);
  Assert.Equal("Name Bulk " + (1000 + 1), entities[1000].Name);
  Assert.Equal("Name Bulk " + (2500 + 1), entities[2500].Name);
  Assert.Equal("Name Bulk " + (4678 + 1), entities[4678].Name);
  Assert.Equal("Name Bulk " + (8888 + 1), entities[8888].Name);
}
```

**Comment by streamBASETdS** [comment]
Yes, I'm using SqlServer; I'll try to create a minimal reproduction-example project, but I didn't succeed just yet (a new minimal project with 8.1.2 showed the same correct behaviour as your test, but bumping my full solution to 8.1.2 shows the errorneous behaviour; I can't share the code of the full solution obviously).

**Comment by MiladNL** [comment]
I am also unable to replicate the issue in a new project.

For our current code, we have reverted to version 8.1.1.

I plan to test it again soon with version 8.1.2 (or newer).

**Comment by streamBASETdS** [comment]
Same nuget 8.1.2 for my test-app

**Comment by darrellzook** [comment]
I recently ran into this issue as well after upgrading from version 8.1.1 to 8.1.2. I used SQL Server Profiler to monitor the generated SQL statements during a bulk insert. Version 8.1.1 generates the following query to return the new list of generated identity values after executing the MERGE statement:

```
SELECT [v].[Id], <other_columns>
FROM (
    SELECT [Id], <other_columns> FROM [dbo].[tableTemp3cd6fbecOutput] WHERE [Id] IS NOT NULL
) AS [v]
ORDER BY [v].[Id]
```

Version 8.1.2 instead generates the following query:

`SELECT [Id], <other_columns> FROM [dbo].[tableTemp91786c7fOutput] WHERE [Id] IS NOT NULL`

Since version 8.1.2 does not produce the wrapping SELECT that orders by the identity column (Id), the results come back in an arbitrary order, which then updates the records with incorrect Id values.

**Comment by borisdj** [comment]
New versions with latest source are published v 8.1.3 for EF8 and v 9.0.1 for EF9.
Make a test for the issue with these versions.

**Comment by marcelwenka** [comment]
Recently, I realised I cannot upgrade to .NET 10 and EntityFrameworkCore 10.0.0 since I'm still stuck at EFCore.BulkExtensions pre 8.1.2.

Since the issues regarding regressions in SetOuputIdentity in BulkInsert were still open I decided to verify myself if the problem has been resolved. I managed to produce a very simple test. Unfortunately, it's quite frail. Sometimes I get multiple _errors_ in a row, sometimes I get a thousand working test cases. Hence it doesn't really make sense to make it into a test case.

If anyone is interested and wants to play around here's the code:

```
using EFCore.BulkExtensions;
using Microsoft.EntityFrameworkCore;

var connectionString = "Server=MyServer;Database=Test;TrustServerCertificate=True;Integrated Security=True"; 
var optionsBuilder = new DbContextOptionsBuilder<TestDbContext>();
optionsBuilder.UseSqlServer(connectionString);
var context = new TestDbContext(optionsBuilder.Options);

var bulkConfig = new BulkConfig
{
    PreserveInsertOrder = true,
    SetOutputIdentity = true
};

var iterations = 10_000;
var recordsCount = 10_000;
for (var i = 1; i <= iterations; i++)
{
    if (i % 100 == 0)
        Console.WriteLine($"{i}/{iterations}");
    await context.TestEntities.ExecuteDeleteAsync();
    var entities = Enumerable.Range(1, recordsCount).Select(x => new TestEntity { Value = x }).ToList();
    await context.BulkInsertAsync(entities, bulkConfig);
    if (entities.Select(x => x.Id).Distinct().Count() != recordsCount)
        Console.WriteLine($"{i}/{iterations} Ids are not unique!");
    if (entities.OrderBy(x => x.Id).Zip(Enumerable.Range(1, recordsCount)).Any(x => x.First.Value != x.Second))
        Console.WriteLine($"{i}/{iterations} Values differ - entities were not inserted in the same order!");
}

public record TestEntity
{
    public int Id { get; set; }
    public int Value { get; set; }
}

public class TestDbContext : DbContext
{
    public TestDbContext(DbContextOptions<TestDbContext> options) : base(options) { }
    public DbSet<TestEntity> TestEntities { get; set; }
}
```

The _Ids are not unique_ console log is never hit. In a different issue someone said they get all Ids equal to 0 even when using SetOutputIdentity. However, the _Values differ - entities were not inserted in the same order!_ does occasionally show up. The test is simple - I create a list of 10 thousand entities with subsequent Values (the entity has two fields: _Id_ and _Value_). If after bulk inserting the order of Ids doesn't correspond to the order of Values it means a _mix-up_ has taken place.

As mentioned before, rerunning the test multiple times might be necessary to encounter the erroneous behaviour. I created a new Test database locally and drop and recreate the table between every few runs.

Since the test is so _flimsy_ I wasn't able to narrow down the version were the issue started and when it was resolved. However, I was able to reproduce the error on multiple runs on the 8.1.2 version and not once on the 10.0.1.

Unfortunately, I haven't initially found this issue, maybe then I would conclude testing earlier. I confirmed on 8.1.2 the SQL to retrieve the data is
```
SELECT [Id], [Value] FROM [dbo].[TestEntitiesTemp4a8fd353Output] WHERE [Id] IS NOT NULL
```
while on 10.0.1 it's
```
SELECT [b].[Id], [b].[Value]
FROM (
    SELECT [Id], [Value] FROM [dbo].[TestEntitiesTempf99502b6Output] WHERE [Id] IS NOT NULL
) AS [b]
ORDER BY [b].[Id]
```

Right now, I believe the issue is caused by SQL usually returning results in the insertion order (or the Id order), but occasionally returning them in a random order, which as we know is not guaranteed and is hard to consistently reproduce. I consider the issue to be resolved as the SQL returned to the proper ordered retrieval and extensive testing on the new version hasn't returned erroneous data.

Unfortunately, only one counterexample is needed to disprove a theory while finding a billion working examples with no counterexamples still doesn't prove the theory's validity so I cannot 100% say it's working correctly now.
